src/gt4py/stencil_object.py

```diff
@@ -277,7 +316,7 @@ def _validate_args(self, field_args, param_args, domain, origin) -> None:
                         f"Origin for field {name} too small. Must be at least {min_origin}, is {field_domain_origin}"
                     )
 
-                spatial_domain = domain.filter_mask(field_domain_mask)
+                spatial_domain = Shape(domain).filter_mask(field_domain_mask)
```

Comment:
Domain was originally typed as a tuple and later modified. It's obvious for humans that after the try-block it cannot be anything else than a `Shape` but not for mypy. The 100% right coding here would be to either use a different variable name for the domain parameter and the _'shapified'_ domain, or to use a user-defined typeguard (https://www.python.org/dev/peps/pep-0647/).

In any case, the current fix should use `typing.cast()` instead of creating a new Shape object
```suggestion
                spatial_domain = typing.cast(Shape, domain).filter_mask(field_domain_mask)
```